Fix Dimensions window values on Android < 15

zoontek opened the pull request:
## Summary:

This PR (initially created for edge-to-edge opt-in support, rebased multiple times) fixes the `Dimensions` API `window` values on Android < 15, when edge-to-edge is enabled.

Currently the window height doesn't include the status and navigation bar heights (but it does on Android >= 15):

<img width="300" alt="Screenshot 2025-06-27 at 16 23 02" src="https://github.com/user-attachments/assets/c7d11334-9298-4f7f-a75c-590df8cc2d8a" />

Using `WindowMetricsCalculator` from AndroidX:

<img width="300" alt="Screenshot 2025-06-27 at 16 34 01" src="https://github.com/user-attachments/assets/7a4e3dc7-a83b-421b-8f6d-fd1344f5fe81" />

Fixes https://github.com/facebook/react-native/issues/47080

## Changelog:

[Android] [Fixed] Fix `Dimensions` `window` values on Android < 15 when edge-to-edge is enabled

## Test Plan:

Run the example app on an Android < 15 device.

janicduplessis commented:
Why would StatusBar backgroundColor prop have no effect in edge to edge? I have cases where I use a semi transparent background in edge to edge mode.

zoontek commented:
@janicduplessis Setting status bar color (`window.setStatusBarColor`) is deprecated and has no effect on Android 15+ (when targetting SDK35) - [source](https://developer.android.com/about/versions/15/behavior-changes-15#edge-to-edge).

> [setStatusBarColor](https://developer.android.com/reference/android/view/Window#setStatusBarColor(int)) and R.attr#statusBarColor are deprecated and have no effect on Android 15.

> Deprecated APIs
> …
>   The following APIs are deprecated and disabled:
>   …
>   [Window#setStatusBarColor](https://developer.android.com/reference/android/view/Window#setStatusBarColor(int))

alanleedev commented:
@zoontek Would there be a way to avoid adding `isEdgeToEdgeEnabled` to `ReactHost`?

zoontek commented:
@alanleedev Would you be OK in using `BuildConfig.IS_EDGE_TO_EDGE_ENABLED` directly?

grahammendick commented:
To enable edge-to-edge in SDK < 35 [Android’s recommendation](https://developer.android.com/develop/ui/views/layout/edge-to-edge#enable-edge-to-edge-display) is to [call enableEdgeToEdge on ComponentActivity](https://developer.android.com/reference/androidx/activity/ComponentActivity#(androidx.activity.ComponentActivity).enableEdgeToEdge(androidx.activity.SystemBarStyle,androidx.activity.SystemBarStyle)). Why isn’t that the approach taken here? And shouldn’t `Appearance.isEdgeToEdge` return true if the Android recommended approach is used (even if that isn’t the approach used by React Native)?

zoontek commented:
@grahammendick Because `Window.enableEdgeToEdge` is also used for `Modal`, which extends `Dialog`. The code is based on [the AndroidX implementation](https://github.com/androidx/androidx/blob/androidx-main/activity/activity/src/main/java/androidx/activity/EdgeToEdge.kt), just not tied to `ComponentActivity`.

grahammendick commented:
@zoontek Most people that use React Native don't use the React Native `Modal` component. So they can turn on edge-to-edge by [calling enableEdgeToEdge](https://developer.android.com/reference/androidx/activity/ComponentActivity#(androidx.activity.ComponentActivity).enableEdgeToEdge(androidx.activity.SystemBarStyle,androidx.activity.SystemBarStyle)). Wouldn’t it be odd if `Appearance.isEdgeToEdge` returned false in that case? Maybe it would be better to work out if edge-to-edge is enabled, for example, by listening for windows insets instead? Then React Native can make its `Modal` edge-to-edge if insets are applied?

grahammendick commented:
> Android doesn't provide a way to properly detect if edge-to-edge is enabled

React Native can add a `OnApplyWindowInsets` listener to the root view. If any top/left/right/bottom inset is > 0 then edge-to-edge is enabled. That should work no matter how edge-to-edge is enabled

zoontek commented:
@grahammendick Even with that, you still need to have this function in RN to apply edge-to-edge on `Modal`. That's why the extra dependency isn't worth it.

cipolleschi commented:
I'm not really the right person to review this as I don't have that extensive knowledge on Android.
I just want to mention that @cortinico is on PTO for a couple of weeks, and he probably have some opinions on this proposal, so I'd kindly ask for a bit of patience and to wait for him to come back. 🙏

kirillzyusko commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/view/WindowUtil.kt:
```diff
@@ -56,8 +56,8 @@ private fun Window.statusBarHide() {
-private fun Window.statusBarShow() {
-  if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+private fun Window.statusBarShow(isEdgeToEdge: Boolean) {
+  if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && !isEdgeToEdge) {
```
Always was hesitant to ask, but now I feel like it's a right time 🙂 

From what I remember `StatusBar` management uses some deprecated methods for managing its properties. As a result it may break some animations driven by ` WindowInsetsAnimationCompat.Callback`.

To overcome this problem I had to create own [module](https://github.com/kirillzyusko/react-native-keyboard-controller/blob/main/android/src/main/java/com/reactnativekeyboardcontroller/modules/StatusBarManagerCompatModuleImpl.kt) based on `WindowInsetsControllerCompat` class and monkey-patch JS module afterward.

This is directly not related to this PR, but I'd like to discuss a possibility to rewrite StatusBar management to a new/modern API 🙌 (whether FB team is open to it and if not then what blocks/prevents that).

kirillzyusko commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/view/WindowUtil.kt:
```diff
@@ -67,39 +67,39 @@ private fun Window.statusBarShow() {
-          else -> Color.argb(0x80, 0x1b, 0x1b, 0x1b)
-        }
+  if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+    isStatusBarContrastEnforced = false
```
Also just a random thought in my head...

Would it be possible to expose a new component (`NavigationBar`, similar to `StatusBar`) and add a possibility to customize properties of NavigationBar (backgroundColor/translucent/style (dark/light))?

I know that there are libraries `react-native-bars` that were aiming to deliver this API. But since we add an ability to toggle `edge-to-edge` mode, maybe it would be reasonable to add other components as well, so that RN will have a full feature set of component out of the box?

zoontek replied:
For me, having a `NavigationBar` component is a bit out-of-scope as a lot of these props will have no effect. `react-native-edge-to-edge` has a `SystemBars` component, which doesn't uses deprecated APIs and align to what is enforced on Android 15. And only `hidden` matters for navigation bar.

- `backgroundColor` -> will have no effect with gesture navigation, only with buttons navigation
- `barStyle` -> same, as gesture navigation style is automatic and based on contrast

For those two to be usable, we also need to set `[setStatusBarContrastEnforced](https://developer.android.com/reference/android/view/Window#setStatusBarContrastEnforced(boolean))` to `false`, which doesn't align with what's enforced in Android 15 / what occurs when AndroidX core`enableEdgeToEdge` is used.

They keep a semi opaque nav bar (white or black, dependending on the color scheme) for buttons nav, transparent for gesture nav. That avoids the additional burden to handle nav bar color / style only for Android users with buttons navigation on (not even the default). I think it's not worth the extra work, and the Google choice is probably the good one here.

zoontek commented on packages/react-native/Libraries/Utilities/Appearance.js:
```diff
@@ -109,6 +113,21 @@ export function setColorScheme(colorScheme: ?ColorSchemeName): void {
+  const state = getState();
+  const {NativeAppearance} = state;
+  if (NativeAppearance != null) {
+    if (state.edgeToEdge == null) {
```
It's just to please the type system, [copied from the line above](https://github.com/facebook/react-native/blob/e101925cd1c0f3428a82392cf1612f1c40c8f79d/packages/react-native/Libraries/Utilities/Appearance.js#L92).

alanleedev commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactHost.kt:
```diff
@@ -48,6 +48,8 @@ public interface ReactHost {
   /** Routes memory pressure events to interested components */
   public val memoryPressureRouter: MemoryPressureRouter
 
+  public fun isEdgeToEdgeEnabled(): Boolean
```
I have some concern with adding this field to ReactHost as this more of a UI configuration that will eventually go away as edge-to-edge will become the default for Android and as we raise the minSdk. Doesn't seem like it is something we should keep in ReactHost.

zoontek replied:
I agree, just tried to copy the current code structure, but it's not optimal. What would you think about removing all of this and just have an `EdgeToEdgeUtils.kt` somewhere?

```kotlin
object EdgeToEdgeUtils {
  fun isEnabled(): Boolean {
    return BuildConfig.IS_EDGE_TO_EDGE_ENABLED
  }
}
```

kkafar commented:
Hey, I want to rekindle the discussion here. Is progress here blocked & we settled with putting source of truth for edge-to-edge in third-party-library or is it still planned to land this proposal eventually in core?

kkafar commented:
Thanks for bringing me up to speed

zoontek commented:
@alanleedev I rebased the pull request against main, removed the opt-out (as it will be enforce on Android 16 / targetSdkVersion 36, there's no need to work with an intermediate RN version with opt-in).

This currently does:

- [x] Apply edge-to-edge on all `Window`: `MainActivity` & `Modal`
- [x] Disable `StatusBar` `translucent` & `backgroundColor`
- [x] Disable `Modal` `statusBarTranslucent` & `navigationBarTranslucent`
- [x] Fix `Dimensions` (`window` values)
- [x] Fix `StatusBar` `hidden` using [immersive mode](https://developer.android.com/develop/ui/views/layout/immersive) (remove `setDecorFitsSystemWindows(true)`)

The rest of the list is optional, as:

- [ ] Add a component to control the navigation bar 👉 Not needed as the navigation bar will be semi-opaque in buttons mode (we keep `isNavigationBarContrastEnforced` at `true`)
- [ ] Allow `StatusBar` / `NavigationBar` to control `Modal` system UI 👉 `ReactModalHostView` `updateSystemAppearance` (the current method) is not as fancy, but might be enough
- [ ] Implement `SafeAreaView` on Android 👉 we can use `react-native-safe-area-context`
- [ ] Implement [modern Keyboard API](https://developer.android.com/develop/ui/views/layout/sw-keyboard) on android 👉 we can use `react-native-keyboard-controller`

zoontek commented:
@alanleedev Than I think I can rollback a few things, but keep this behavior enforced when targetSdk is >= 36, instead of a flag or an opt-in (using `context.getApplicationInfo().targetSdkVersion`)

zoontek commented:
@alanleedev Given all of that, keeping the opt-in flag is the better option. I will rollback this part when I'm home.

zoontek commented:
@alanleedev I added back the `edgeToEdgeEnabled` gradle setting + a `IS_EDGE_TO_EDGE_ENABLED` constant in the `Appearance` module (for the `react-native-is-edge-to-edge` detection lib). So, in the current state, we have:

- Edge-to-edge on all Window: `MainActivity` & `Modal` instances
- No-op on `StatusBar` `translucent` & `backgroundColor`
- Enforced `Modal` `statusBarTranslucent` & `navigationBarTranslucent`
- Fixed `Dimensions` API (`window` values)
- Fixed `hidden` on `StatusBar` [to use immersive mode](https://developer.android.com/develop/ui/views/layout/immersive) (avoid `setDecorFitsSystemWindows(true)`)

The navigation bar behavior is similar to [the one enforced on Android 15](https://developer.android.com/about/versions/15/behavior-changes-15#edge-to-edge):
- On Android < 8, the navigation bar (buttons) is semi-opaque black
- On Android < 10, the navigation bar (buttons) is semi-opaque white
- On Android >= 10
  - The navigation bar (when gestures are used) is transparent. The bar style is automatic, based on contrast
  - The navigation bar (when buttons are used) is semi opaque black or white, based on the system setting (light or dark mode)

We *could* add a `NavigationBar` component later, but the PR is reviewable right now.

alanleedev commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java:
```diff
@@ -121,6 +123,9 @@ public void onCreate(Bundle savedInstanceState) {
           String mainComponentName = getMainComponentName();
           final Bundle launchOptions = composeLaunchOptions();
+          if (ReactBuildConfig.IS_EDGE_TO_EDGE_ENABLED && mActivity != null) {
+            WindowUtilKt.enableEdgeToEdge(mActivity.getWindow());
```
@zoontek 
I know it probably is bit difficult to directly call this from here but have you considered using AndroidX [ComponentActivity.enableEdgeToEdge()](https://developer.android.com/reference/kotlin/androidx/activity/ComponentActivity#(androidx.activity.ComponentActivity).enableEdgeToEdge(androidx.activity.SystemBarStyle,androidx.activity.SystemBarStyle)) as it also seems to have additional logic around handling status bar and navigation bar ([source code](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:activity/activity/src/main/java/androidx/activity/EdgeToEdge.kt))?

zoontek replied:
That's would require migrating `Activity` to `ComponentActivity` in `ReactActivityDelegate` (and everywhere needed in the codebase), a bigger change. Are you good with this?

Also, as `enableEdgeToEdge` is not available on `Dialog`, `WindowUtilKt.enableEdgeToEdge` will still be needed for `Modal`. That's why I reused it.

alanleedev replied:
Seems like you already answer this here; https://github.com/facebook/react-native/pull/47554#issuecomment-2473017391

alanleedev commented on packages/react-native/src/private/specs_DEPRECATED/modules/NativeAppearance.js:
```diff
@@ -19,6 +19,9 @@ export type AppearancePreferences = {
 };
 
 export interface Spec extends TurboModule {
+  +getConstants: () => {
+    +IS_EDGE_TO_EDGE_ENABLED: boolean,
+  };
```
@zoontek Have you considered adding this to Dimensions? Appearance is more about themes and Dimensions is more size related and edge-to-edge seems closer to size.

alanleedev commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactHostImpl.kt:
```diff
@@ -616,6 +617,10 @@ public class ReactHostImpl(
   @ThreadConfined(ThreadConfined.UI)
   override fun onConfigurationChanged(context: Context) {
+    if (ReactBuildConfig.IS_EDGE_TO_EDGE_ENABLED) {
+      UiThreadUtil.runOnUiThread { currentActivity?.window?.enableEdgeToEdge() }
```
@zoontek What happens if we leave this out? Or why is this needed?

zoontek replied:
This was a defensive mechanism against a few libraries that disable edge-to-edge by calling `setDecorFitsSystemWindows(true)` in some cases.

Those cases has been tackled with mutliple PRs accross the ecosystem, should be good to remove this now.

alanleedev commented:
@zoontek I think it would help if we can eventually break the PR into smaller parts (not necessarily in this order).
- ContextUtils.isDarkMode()
- IS_EDGE_TO_EDGE_ENABLED 
- DisplayMetricsHolder
- RNTesterActivity
- edge-to-edge in Modal

> - Fixed Dimensions API (window values).   

It would be good to write up more details on what was fixed.
Also there is potential work to migrate the deprecated getRealMetrics in DisplayMetricsHolder by @mojavad

zoontek commented on packages/react-native/React/CoreModules/RCTAppearance.mm:
```diff
@@ -92,6 +93,13 @@ - (instancetype)init
+- (void)initialize
+{
+  _constants = facebook::react::typedConstants<JS::NativeAppearance::Constants>({
+      .IS_EDGE_TO_EDGE_ENABLED = true
```
The value is not exposed in the public API, I plan to use it in [`react-native-is-edge-to-edge`](https://github.com/zoontek/react-native-edge-to-edge/tree/main/react-native-is-edge-to-edge) and the value check is only done on Android.

It's only declared on iOS because of codegen (similar to https://github.com/facebook/react-native/blob/v0.80.0-rc.5/packages/react-native/React/CoreModules/RCTStatusBarManager.mm#L82)

zoontek commented:
@alanleedev Here is the next PR: https://github.com/facebook/react-native/pull/52088
The last one will be to fix the `Dimensions` API (`window` values)

alanleedev commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/view/WindowUtil.kt:
```diff
@@ -13,9 +13,20 @@ import android.view.Window
 
+// The light scrim color used in the platform API 29+
+// https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/com/android/internal/policy/DecorView.java;drc=6ef0f022c333385dba2c294e35b8de544455bf19;l=142
+internal val LightNavigationBarColor = Color.argb(0xe6, 0xFF, 0xFF, 0xFF)
```
does this need to be `internal` ?

zoontek replied:
![Screenshot 2025-06-18 at 09 22 48](https://github.com/user-attachments/assets/78c0105a-6ceb-4a15-ab25-696691a2d26a)

Either it's `internal` or public, and I'm not sure that's what we want / need (similar to [AndroidX](https://github.com/androidx/androidx/blob/androidx-main/activity/activity/src/main/java/androidx/activity/EdgeToEdge.kt#L37))

cortinico commented:
> @cortinico Done ✅
> 
> I rebased it for the last time / the last fix: the `Dimensions` `window` values on Android < 15. Let me know if something is missing.
> 
> PS: I added AndroidX window library as the code could be [quite big to embed](https://github.com/androidx/androidx/blob/androidx-main/window/window/src/main/java/androidx/window/layout/util/BoundsHelper.kt?rgh-link-date=2025-06-27T14%3A49%3A57Z).

Great also is this PR going to fix https://github.com/facebook/react-native/issues/47080 ?

cortinico commented:
> @cortinico @alanleedev Are you interested in a `NavigationBar` component directly in core? Or should it stays third-party?

IMHO it should stay third-party (as `StatusBar` should have probably been a 3p component as well)

alanleedev commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/DisplayMetricsHolder.kt:
```diff
@@ -62,16 +64,30 @@ public object DisplayMetricsHolder {
+    windowDisplayMetrics.setTo(displayMetrics)
     screenDisplayMetrics.setTo(displayMetrics)
+
+    if (isEdgeToEdgeFeatureFlagOn) {
```
Does this only change the behavior when only `isEdgeToEdgeFeatureFlagOn` is set?

zoontek commented:
> @zoontek I'm running into circular dependency issues with internal BUCK builds. May bit tricky to resolve. Will update later.

@alanleedev Sure, let me know

cortinico commented:
@zoontek we had to revert this because of integration issues

cortinico commented:
> @cortinico Is there something that could be done?

We're looking inot it. We'll keep you in the loop if there is anything needed on your end 👍

cortinico commented:
Hey @zoontek 
I had to revert this change on the 0.81 release branch because it's making RNTester on Android 7.0 instacrash with:

```
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: Exception in native call
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: java.lang.IllegalArgumentException: Context com.facebook.react.runtime.BridgelessReactContext@97f69a1 is not a UiContext
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at androidx.window.layout.util.ContextCompatHelper.unwrapUiContext$window_release(ContextCompatHelper.kt:52)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at androidx.window.layout.util.WindowMetricsCompatHelperBaseImpl.currentWindowMetrics(WindowMetricsCompatHelper.kt:84)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at androidx.window.layout.WindowMetricsCalculatorCompat.computeCurrentWindowMetrics(WindowMetricsCalculatorCompat.kt:39)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.uimanager.DisplayMetricsHolder.initDisplayMetrics(DisplayMetricsHolder.kt:74)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.uimanager.DisplayMetricsHolder.initDisplayMetricsIfNotInitialized(DisplayMetricsHolder.kt:61)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.ReactInstance.<init>(ReactInstance.kt:243)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.ReactHostImpl.getOrCreateReactInstanceTask$lambda$41$lambda$38(ReactHostImpl.kt:919)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.ReactHostImpl.$r8$lambda$LQY77bFECGrnkUqd7Ku55sJ721A(ReactHostImpl.kt)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.ReactHostImpl$$ExternalSyntheticLambda22.then(D8$$SyntheticClass:0)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task$Companion.completeImmediately$lambda$3(Task.kt:356)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task$Companion.$r8$lambda$b7MTC0ufkn_2TjF4BGcNqGR9qK0(Task.kt)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task$Companion$$ExternalSyntheticLambda3.run(D8$$SyntheticClass:0)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Executors$ImmediateExecutor.execute(Executors.kt:41)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task$Companion.completeImmediately(Task.kt:354)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task$Companion.access$completeImmediately(Task.kt:250)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task.continueWith(Task.kt:132)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task.continueWith$default(Task.kt:117)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task.onSuccess$lambda$12(Task.kt:174)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task.$r8$lambda$l0wwo5AvbHxlwfI1R7CH_codxoo(Task.kt)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task$$ExternalSyntheticLambda2.then(D8$$SyntheticClass:0)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task$Companion.completeAfterTask$lambda$5(Task.kt:390)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task$Companion.$r8$lambda$IPpysnehFwANEiW061YzapkMdoE(Task.kt)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.internal.bolts.Task$Companion$$ExternalSyntheticLambda4.run(D8$$SyntheticClass:0)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at java.lang.Thread.run(Thread.java:761)
```

Could you look into it?

cc @alanleedev for visibility

alanleedev commented:
@zoontek 
Seem like the crash log shows it being called from ReactInstance and not ReactRootView. Were you able to verify the crash on Android 17 before the fix?

> 07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.uimanager.DisplayMetricsHolder.initDisplayMetricsIfNotInitialized(DisplayMetricsHolder.kt:61)
> 07-14 14:38:13.572  3999  4019 E unknown:ReactNative: 	at com.facebook.react.runtime.ReactInstance.<init>(ReactInstance.kt:243)

zoontek commented:
@alanleedev Yes, tried it on an Android 7 emulator (check the video). Isn't the `ReactInstance` what is returned by `getContext().getApplicationContext()`. Does the issue still happens in E2E?

zoontek commented:
@alanleedev When I tried it with `getContext().getApplicationContext()`, I successfully reproduce the crash: first at app start, and once the first occurence (line 868) has been replaced, at app orientation change (line 991).

By passing UI context (`getContext().getApplicationContext()` -> `getContext()`), no more crashed on my side.

alanleedev commented:
@zoontek That is weird, it does not repro for me on my emulator running on ARM based Mac. At least, I don't see any issues with the code fix. I guess we can try to merge and see if we run into any issues in the release branch.

cortinico commented:
Hey @zoontek 

Sadly this is still crashing, on **release only** and on Android 7 only with the following stacktrace:

```
2025-07-18 17:02:19.971  5958-5958  AndroidRuntime          com.facebook.react.uiapp             E  FATAL EXCEPTION: main
                                                                                                    Process: com.facebook.react.uiapp, PID: 5958
                                                                                                    java.lang.IllegalArgumentException: Context com.facebook.react.runtime.BridgelessReactContext@a6a81db is not a UiContext
                                                                                                    	at androidx.window.layout.util.ContextCompatHelper.unwrapUiContext$window_release(ContextCompatHelper.kt:52)
                                                                                                    	at androidx.window.layout.util.WindowMetricsCompatHelperBaseImpl.currentWindowMetrics(WindowMetricsCompatHelper.kt:84)
                                                                                                    	at androidx.window.layout.WindowMetricsCalculatorCompat.computeCurrentWindowMetrics(WindowMetricsCalculatorCompat.kt:39)
                                                                                                    	at com.facebook.react.uimanager.DisplayMetricsHolder.initDisplayMetrics(DisplayMetricsHolder.kt:74)
                                                                                                    	at com.facebook.react.uimanager.DisplayMetricsHolder.initDisplayMetricsIfNotInitialized(DisplayMetricsHolder.kt:61)
                                                                                                    	at com.facebook.react.runtime.ReactInstance.<init>(ReactInstance.kt:243)
                                                                                                    	at com.facebook.react.runtime.ReactHostImpl.getOrCreateReactInstanceTask$lambda$41$lambda$37(ReactHostImpl.kt:919)
                                                                                                    	at com.facebook.react.runtime.ReactHostImpl.$r8$lambda$NwQSP6waC5OEu_tQNW6qI0fv5jo(ReactHostImpl.kt)
                                                                                                    	at com.facebook.react.runtime.ReactHostImpl$$ExternalSyntheticLambda11.then(D8$$SyntheticClass:0)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task$Companion.completeImmediately$lambda$3(Task.kt:356)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task$Companion.$r8$lambda$b7MTC0ufkn_2TjF4BGcNqGR9qK0(Task.kt)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task$Companion$$ExternalSyntheticLambda3.run(D8$$SyntheticClass:0)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Executors$ImmediateExecutor.execute(Executors.kt:41)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task$Companion.completeImmediately(Task.kt:354)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task$Companion.access$completeImmediately(Task.kt:250)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task.continueWith(Task.kt:132)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task.continueWith$default(Task.kt:117)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task.onSuccess$lambda$12(Task.kt:174)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task.$r8$lambda$l0wwo5AvbHxlwfI1R7CH_codxoo(Task.kt)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task$$ExternalSyntheticLambda2.then(D8$$SyntheticClass:0)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task$Companion.completeAfterTask$lambda$5(Task.kt:390)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task$Companion.$r8$lambda$IPpysnehFwANEiW061YzapkMdoE(Task.kt)
                                                                                                    	at com.facebook.react.runtime.internal.bolts.Task$Companion$$ExternalSyntheticLambda4.run(D8$$SyntheticClass:0)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
                                                                                                    	at java.lang.Thread.run(Thread.java:761)
```

cortinico commented:
cc @alanleedev

zoontek commented:
@cortinico @alanleedev There's probably a race condition between `initDisplayMetricsIfNotInitialized` usages. If the first call is performed with an UI context, all good. If it's not, it fails.

Indeed, we need to rethink the approach. For me, `initDisplayMetrics` argument should be replaced by `activity: Activity` and usage must be fixed. Let me have a look, to see if this can be done quickly.

zoontek commented:
@cortinico @alanleedev I've been able to reproduce the issue on Android 7, in **release** mode only (this doesn't occur in dev mode).

I switched `initDisplayMetricsIfNotInitialized` and `initDisplayMetrics` signatures to either accept a `ReactContext?` or a `Activity?` instead, and guarantee init on the initial UI context passed.
I ran it on Android 7 debug and release mode, and various post Android 10 versions, debug and release mode, no crash.

The fix is here: https://github.com/zoontek/react-native/pull/1/files

zoontek commented:
@cortinico Sure thing, here it is: https://github.com/facebook/react-native/pull/52738

Tartarus2014 commented:
Can React-Native force open Android's edgetoedge?
